x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Authenticator.java

```diff
@@ -93,15 +86,52 @@ class Context implements Closeable {
         private final AuthenticationService.AuditableRequest request;
         private final User fallbackUser;
         private final boolean allowAnonymous;
+        // {@code true}, if {@code Authenticator}s should first be tried in order to extract the credentials token from the thread context
+        // otherwise, the tokens must explicitly be passed in, see {@code authenticationTokens}
+        private final boolean extractCredentials;
         private final Realms realms;
-        private final List<AuthenticationToken> authenticationTokens = new ArrayList<>();
+        // {@code Authenticator}s are consulted in order (see {@code AuthenticatorChain}),
+        // where each is given the chance to first extract some token, and then to verify it.
+        // If token verification fails in some particular way (i.e. {@code AuthenticationResult.Status.CONTINUE}),
+        // the next {@code Authenticator} is tried. The extracted tokens are appended to this list.
```

Comment:
> I don't really get why we preserve the extracted tokens across Authenticators

Yes, I think we don't need to preserve them, as you describe.
It's great we discuss it, but I'm reluctant to handle any of it under this PR.

It might be that some global view of all the extracted tokens can be used in a couple of ancilliary places, such as
https://github.com/elastic/elasticsearch/blob/c10f94f3ace1677eb3d3800f8ef399578a214541/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java#L269
and https://github.com/elastic/elasticsearch/blob/c10f94f3ace1677eb3d3800f8ef399578a214541/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java#L373 .

If we preserve the global view, I agree that a _list_ is a poor choice as a data structure, IMO preferable would be a mapping from authenticator->extracted_token.

Reply:
> but I'm reluctant to handle any of it under this PR.

++ it is outside the scope of the PR. thanks for helping to confirm i was not missing something.